et_replay/comm/profiler_trace_analysis.py

```diff
@@ -0,0 +1,252 @@
+    'gather'            : (lambda n: (n - 1)/n),
+    'broadcast'         : (lambda n: 1),
+    'send'              : (lambda n: 1),
+    'recv'              : (lambda n: 1),
+}
+
+# map collective name of event to key string for bw calculation
+_collname_to_busbw_corr_factor_func: Dict[str, Callable[[int], float]] = {
+    'allreduce'             : _busbw_correction_factors_tbl['all_reduce'],
+    'all_gather'            : _busbw_correction_factors_tbl['all_gather'],
+    '_allgather_base'       : _busbw_correction_factors_tbl['all_gather'],
+    'reduce_scatter'        : _busbw_correction_factors_tbl['reduce_scatter'],
+    '_reduce_scatter_base'  : _busbw_correction_factors_tbl['reduce_scatter'],
+    'all_to_all'            : _busbw_correction_factors_tbl['all_to_all'],
+    'all_to_allv'           : _busbw_correction_factors_tbl['all_to_all'],
+    'broadcast'             : _busbw_correction_factors_tbl['broadcast'],
+    'reduce'                : _busbw_correction_factors_tbl['reduce'],
+    'gather'                : _busbw_correction_factors_tbl['gather'],
+    'scatter'               : _busbw_correction_factors_tbl['scatter'],
+    'send'                  : _busbw_correction_factors_tbl['send'],
+    'recv'                  : _busbw_correction_factors_tbl['recv'],
+}
+
+def _get_dict_value(d, k, err_msg):
+    if k not in d:
+        raise ValueError(err_msg)
+    return d.get(k)
+
+def _calculate_event_data_size(evt):
+    return max(evt['args']['In msg nelems'], evt['args']['Out msg nelems']) * _dtype_size_map[evt['args']['dtype']]
+
+def _calculate_algbw(evt: Dict[str, Any]) -> float:
+    duration_us = _get_dict_value(evt, 'dur', f'Missing "dur" in event: {evt}')
+    total_bytes = _calculate_event_data_size(evt)
+
+    # NCCL tests use 1024^3 to convert B to GB (but not 1e9)
+    # https://github.com/NVIDIA/nccl-tests/blob/8dfeab9eb9bdfdf13503e71e1f33e7f8a208b540/src/common.cu#L102
+    # but it uses 1e9 to convert bw from B/s to GB/s
+    # https://github.com/NVIDIA/nccl-tests/blob/8dfeab9eb9bdfdf13503e71e1f33e7f8a208b540/src/all_gather.cu#L41
+    return round((total_bytes / duration_us) / 1e3, 2)
+
+def _get_event_busbw_factor(evt):
+    coll_name = _get_dict_value(evt['args'], 'Collective name', f'Missing "Collective name" in event: {evt}')
+    
+    # barrier is implemented using AllReduce            
+    if coll_name in ['barrier', ]:
+        return 0
+    
+    group_size = _get_dict_value(evt['args'], 'Group size', f'Missing "Group size" in event: {evt}')
+    correction_factor_func = _get_dict_value(_collname_to_busbw_corr_factor_func,
+                                             coll_name, 
+                                             f'Unsupported collective op for busbw calculation: {coll_name}')
+    
+    return correction_factor_func(group_size)
+
+def calculate_bw_(trace_data):
+    nccl_events = [i for i in trace_data['traceEvents'] if i.get('cat', '') == 'kernel' and i['name'].startswith('ncclDevKernel_')]
```

Comment:
to save memory consider using generator
nccl_events = (e for e in trace_data['traceEvents'] if e.get('cat', '') == 'kernel' and e['name'].startswith('ncclDevKernel_'))